Create hash-table.h

ProfessorLayton322 opened the pull request:
Код хэш таблицы
Фёдор Быков 201
https://official.contest.yandex.ru/contest/24924/run-report/49046502/

ProfessorLayton322 committed:
Create hash-table.h

Код хэш таблицы

Commit:
1. Comments
2. Description
3. Minor style changes

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+/*
+Hash table.
+Using chain method.
+It doubles the table size if needed to expand.
```
А за счёт чего линейное итерирование? Описание недостаточное

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+Hash table.
+Using chain method.
+It doubles the table size if needed to expand.
+Check https://pastebin.com/esx8ckYw for more.
```
Зачем это вставлено через pastebin?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+Using chain method.
+It doubles the table size if needed to expand.
+Check https://pastebin.com/esx8ckYw for more.
+*/
```
Не хватает ссылки на полноценное описание реализации, например на википедию

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+    typedef typename std::list<std::pair<const KeyType, ValueType> >::iterator iterator;
+    typedef typename std::list<std::pair<const KeyType, ValueType> >::const_iterator const_iterator;
+
+  private:
```
приватные поля идут в конце, никак не до публичных методов

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+    //Doubles the capacity and rebuilds the table.
+    void expand() {
+        table.clear();
+        capacity *= 2;
```
Управляющая константа 2 должна быть вынесена в константное поле класса

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+        capacity *= 2;
+        table.resize(capacity);
+        for (auto it = begin(); it != end(); ++it) {
+            int hash = hasher(it->first) % capacity;
```
почему int ?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+    void add(const std::pair<const KeyType, ValueType>& insertValue) {
+        sz++;
+        int hash = hasher(insertValue.first) % capacity;
+        if (size() == capacity) {
```
Действие "проверь нормальный ли размер, и, если надо, увеличь" явно нужно вынести в отдельный метод

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    //Returns size.
+    size_t size() const {
```
а почему стиль именования публичных методов не отличается от стиля именования приватных полей?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+                                          const inputIterator inputEnd,
+                                          const Hash& _hasher = Hash()):
+        hasher(_hasher),
+        capacity(std::distance(inputBegin, inputEnd)),
```
такое решение не будет работать с однопроходными итераторами

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+                                          const Hash& _hasher = Hash()):
+        hasher(_hasher),
+        capacity(std::distance(inputBegin, inputEnd)),
+        table(std::distance(inputBegin, inputEnd)),
```
и почему мы сразу делаем таблицу с таким capacity, что после заполнения её придётся перестраивать?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+            capacity(input.size()),
+            table(input.size()),
+            sz(0) {
+            if (input.begin() == input.end()) {
```
выравнивание

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+                table.resize(1);
+                return;
+            }
+        for (auto it = input.begin(); it != input.end(); ++it) {
```
почему не range-based for?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+            int hash = hasher(it->first) % capacity;
+            bool flag = false;
+            for (iterator contentIt : table[hash])
+                if (contentIt->first == it->first)
```
нет фигурных скобок после if

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+            for (iterator contentIt : table[hash])
+                if (contentIt->first == it->first)
+                    flag = true;
+            if (flag)
```
и тут нет скобок

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+        for (auto it = input.begin(); it != input.end(); ++it) {
+            int hash = hasher(it->first) % capacity;
+            bool flag = false;
+            for (iterator contentIt : table[hash])
```
почему не константный итератор?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,273 @@
+                capacity = 1;
+                table.resize(capacity);
+            }
+            for (auto it = content.begin(); it != content.end(); ++it) {
```
опять же можно range-based for

ProfessorLayton322 replied:
Нет нельзя, мне нужен итератор чтобы закинуть его дальше в таблицу

ProfessorLayton322 committed:
Update hash-table.h

All local fields now have _ in their names

ProfessorLayton322 committed:
Update hash-table.h

Description changed

ProfessorLayton322 committed:
Update hash-table.h

2 is now a capacityInflation constexpression

ProfessorLayton322 committed:
Update hash-table.h

Replaxced expand() with checkExpansion(), see comments

ProfessorLayton322 committed:
Update hash-table.h

iterator constructor changed

ProfessorLayton322 committed:
Update hash-table.h

Every constructor now allocates two times more memory than elements given

ProfessorLayton322 committed:
Update hash-table.h

style changes

ProfessorLayton322 committed:
Update hash-table.h

initializer list constructor changed

ProfessorLayton322 committed:
Update hash-table.h

style changes

ProfessorLayton322 committed:
Update hash-table.h

const iterators where needed

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,279 @@
+    void add(const std::pair<const KeyType, ValueType>& insertValue) {
+        _sz++;
+        size_t hash = _hasher(insertValue.first);
+        checkExpansion();
```
очень странно, почему мы сначала увеличиваем счётчик, потом пробуем отмасштабироваться, потом вставляем элемент? Вроде checkExpansion должен быть в конце

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,279 @@
+    //Inserts new <Key, Value> pair. Checks if table already already has this key.
+    void insert(const std::pair<const KeyType, ValueType>& insertValue) {
+        size_t hash = _hasher(insertValue.first) % _capacity;
+        bool flag = false;
```
это дублирование кода find, почему бы не использовать его здесь?

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,279 @@
+
+    //Doubles the capacity and rebuilds the table if size reached capacity
+    void checkExpansion() {
+        if (size() == _capacity) {
```
вроде бы всё равно, но идейно правильнее написать >=

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,279 @@
+            _content.push_back(currentPair);
+            _it++;
+        }
+        for (size_t i = 0; i < my_size; i++)
```
скобки

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,279 @@
+            _capacity = 1;
+        }
+        _table.clear();
+        _table.resize(_capacity);
```
получается, мы сразу создаём переполненную таблицу?

ProfessorLayton322 committed:
Update hash-table.h

Changed checkExpansion() calls

ProfessorLayton322 committed:
Update hash-table.h

Insert method changed (shorter)

ProfessorLayton322 committed:
Update hash-table.h

changed assign operator (allocates more memory), also style changes

glebshp commented on hash-table.h:
```diff
@@ -0,0 +1,272 @@
+    HashMap<KeyType, ValueType, Hash>& operator = (const HashMap<KeyType, ValueType, Hash>& other) {
+        size_t other_size = other.size();
+        size_t my_size = size();
+        auto _it = other.begin();
```
не понятно что тут происходит, либо зарефакторить, либо добавить комменты

ProfessorLayton322 committed:
Update hash-table.h

1. Private method fillTable - fills empty table with content iterators.
2. Refactored assign operator.
3. Minor change to checkExpansion
